File: javascript/script/recursion.js
function factorial (n) {
    if (n === 0 || n === 1) return 1
    return n * factorial(n - 1)
}

function sumArr (arr) {
    if (arr.length === 0) return 0
    return arr[0] + sumArr(arr.slice(1)) 
}

function countListItems (arr) {
    if (arr.length === 0) return 0
    return 1 + countListItems(arr.slice(1))
}

function getMin (arr) {
    if (arr.length === 0) return null
    if (arr.length === 1) return arr[0]
    let left = arr[0]
    let right = getMin(arr.slice(1))
    return left < right ? left : right
}

function binarySearch (arr, num) {
    if (arr.length === 0) return null
    if (arr.length === 1) {
        if (arr[0] == num) return arr[0]
        return null
    }

    let mid = Math.floor(arr.length / 2)
    let guess = arr[mid]
    if (guess ===  num) {
        return guess
    } else if (guess < num) {
        return binarySearch(arr.slice(mid+1), num)
    } else if (guess > num) {
        return binarySearch(arr.slice(0, mid), num)
    }
}

module.exports = {
    factorial,
    sumArr,
    countListItems,
    getMin,
    binarySearch
}

File: python/script/recursion.py
def factorial(x):
    if x == 1:
        return 1
    return x * factorial(x - 1)


def sum_arr(arr):
    if not arr:
        return 0
    return arr[0] + sum_arr(arr[1:])


def count_list_items(arr):
    if not arr:
        return 0
    return 1 + count_list_items(arr[1:])


def get_max(arr):
    if len(arr) < 2:
        return arr[0]
    left = arr[0]
    right = get_max(arr[1:])
    return left if left > right else right


def binary_search(arr, num):
    if len(arr) == 0:
        return None

    if len(arr) == 1:
        return arr[0] if arr[0] == num else None

    mid = len(arr) // 2
    guess = arr[mid]

    if guess == num:
        return guess
    elif guess < num:
        return binary_search(arr[mid + 1 :], num)
    elif guess > num:
        return binary_search(arr[:mid], num)
